src/main/java/com/amazonaws/kinesisvideo/demoapp/activity/WebRtcActivity.java

```diff
@@ -259,6 +266,47 @@ private boolean isValidClient() {
         return client != null && client.isOpen();
     }
 
+    private void handlePendingIceCandidates(String clientId) {
+        // Add any pending ICE candidates from the queue for the client ID
+        Queue<IceCandidate> pendingIceCandidatesQueueByClientId = pendingIceCandidatesMap.get(clientId);
+        while(!pendingIceCandidatesQueueByClientId.isEmpty()) {
+            final IceCandidate iceCandidate = pendingIceCandidatesQueueByClientId.peek();
+            final boolean addIce = localPeer.addIceCandidate(iceCandidate);
+            Log.d(TAG, "Added ice candidate after SDP exchange " + iceCandidate + " " + (addIce ? "Successfully" : "Failed"));
+          pendingIceCandidatesQueueByClientId.remove();
+        }
+        // After sending pending ICE candidates, the client ID's peer connection need not be tracked
+        pendingIceCandidatesMap.remove(clientId);
+        peerConnectionFoundMap.remove(clientId);
+    }
+
+    private void checkAndAddIceCandidate(Event message, IceCandidate iceCandidate) {
+        // if answer/offer is not received, it means peer connection is not found. Hold the received ICE candidates in the map.
+        if(!peerConnectionFoundMap.containsKey(message.getSenderClientId())) {
+            Log.d(TAG, "SDP exchange is not complete. Ice candidate " + iceCandidate + " + added to pending queue");
+            // Add ICE candidate to a queue
+            peerConnectionFoundMap.put(message.getSenderClientId(), false);
```

Comment:
Here, you should not be adding to this map. This map (name should change to reflect clientID->peer connection object) should be added ONLY after you get an answer/offer.

NOTE: The entry should be removed at the end of the session